crates/turbo-tasks-macros-shared/src/expand.rs

```diff
@@ -103,15 +105,22 @@ pub fn expand_fields<
 
 /// Generates a match arm destructuring pattern for the given fields.
 ///
+/// If no `filter_field` function is provided, all fields are included in the
+/// pattern. If a `filter_field` function is provided, only fields for which
+/// the function returns `true` are included in the pattern. If any field is
+/// ignored, a wildcard pattern is added to the end of the pattern, making it
+/// non-exhaustive.
+///
 /// Returns both the capture pattern token stream and the name of the bound
 /// identifiers corresponding to the input fields.
 pub fn generate_destructuring<'a, I: Fn(&Field) -> bool>(
-    fields: impl Iterator<Item = &'a Field>,
-    ignore_field: &I,
+    fields: impl Iterator<Item = &'a Field> + ExactSizeIterator,
+    filter_field: &I,
 ) -> (TokenStream, Vec<TokenStream>) {
+    let fields_len = fields.len();
     let (captures, fields_idents): (Vec<_>, Vec<_>) = fields
+        .filter(|field| filter_field(field))
         .enumerate()
-        .filter(|(_i, field)| !ignore_field(field))
```

Comment:
I think this is a bug. The enumerate is done first so that we can get the index of the item in input. If we filter first, then index 1 could refer to the third field.